external/avro/src/main/scala/org/apache/spark/sql/avro/AvroOptions.scala

```diff
@@ -78,7 +80,17 @@ class AvroOptions(
     parameters
       .get(AvroOptions.ignoreExtensionKey)
       .map(_.toBoolean)
-      .getOrElse(!ignoreFilesWithoutExtension)
+      .map { ignoreExtensionOption =>
+        if (ignoreExtensionOption != !ignoreFilesWithoutExtensionByDefault) {
+          warn(s"The Avro option '${AvroOptions.ignoreExtensionKey}'")
```

Comment:
@MaxGekk, from a cursory look, this warning can be shown for every file:

https://github.com/apache/spark/blob/053dd858d38e6107bc71e0aa3a4954291b74f8c8/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/FilePartitionReaderFactory.scala#L24-L30

https://github.com/apache/spark/blob/053dd858d38e6107bc71e0aa3a4954291b74f8c8/external/avro/src/main/scala/org/apache/spark/sql/v2/avro/AvroPartitionReaderFactory.scala#L61

Do you mind if I ask double check this?

Reply:
@HyukjinKwon I will check that but general thoughts are:
1. The log warning is printed only if an user sets non-default config values
2. I don't think `AvroOptions` should be created per-each file if it is created in current implementation.

Reply:
@HyukjinKwon you are right, it prints warnings per each partition. I have confirmed that by the test:
```scala
  test("count deprecation log events") {
    val partitionNum = 3
    val logAppender = new AppenderSkeleton {
      val loggingEvents = new ArrayBuffer[LoggingEvent]()

      override def append(loggingEvent: LoggingEvent): Unit = loggingEvents.append(loggingEvent)
      override def close(): Unit = {}
      override def requiresLayout(): Boolean = false
    }
    withTempPath { dir =>
      Seq(("a", 1, 2), ("b", 1, 2), ("c", 2, 1), ("d", 2, 1))
        .toDF("value", "p1", "p2")
        .repartition(partitionNum)
        .write
        .format("avro")
        .option("header", true)
        .save(dir.getCanonicalPath)
      withLogAppender(logAppender) {
        val df = spark
          .read
          .format("avro")
          .schema("value STRING, p1 INTEGER, p2 INTEGER")
          .option(AvroOptions.ignoreExtensionKey, false)
          .option("header", true)
          .load(dir.getCanonicalPath)
        df.count()
      }
      val deprecatedEvents = logAppender.loggingEvents
        .map(_.getRenderedMessage)
        .filter(_.contains(AvroOptions.ignoreExtensionKey))
      assert(deprecatedEvents.size === partitionNum)
    }
  }
```

Reply:
When I moved instantiation of AvroOptions out of buildReader():
<img width="857" alt="Screen Shot 2020-01-12 at 20 14 06" src="https://user-images.githubusercontent.com/1580697/72222647-359c3a00-3578-11ea-92d3-0802dbb21b3c.png">
The warning is printed always 2 times. It means `AvroPartitionReaderFactory` is constructed twice.
And both times from https://github.com/apache/spark/blob/053dd858d38e6107bc71e0aa3a4954291b74f8c8/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2ScanExecBase.scala#L60-L66 which is called:
First time from
<img width="860" alt="Screen Shot 2020-01-12 at 20 32 02" src="https://user-images.githubusercontent.com/1580697/72222900-a8a6b000-357a-11ea-8792-2e94532199f0.png">
The second time from:
<img width="656" alt="Screen Shot 2020-01-12 at 20 33 36" src="https://user-images.githubusercontent.com/1580697/72222915-d5f35e00-357a-11ea-9f91-da36c59d62ff.png">

Reply:
It is interesting that rewriting `supportsColumnar` as:
```scala
  override val supportsColumnar: Boolean = {
    val factory = readerFactory
    require(partitions.forall(factory.supportColumnarReads) ||
      !partitions.exists(factory.supportColumnarReads),
      "Cannot mix row-based and columnar input partitions.")

    partitions.exists(factory.supportColumnarReads)
  }
```
does not help too because `DataSourceV2ScanExecBase` is initialized twice from:
First time:
<img width="851" alt="Screen Shot 2020-01-12 at 20 52 49" src="https://user-images.githubusercontent.com/1580697/72223223-84000780-357d-11ea-829e-768d37d3fd1f.png">
Second time in `TreeNode.makeCopy`:
<img width="797" alt="Screen Shot 2020-01-12 at 21 06 40" src="https://user-images.githubusercontent.com/1580697/72223406-73e92780-357f-11ea-9c18-86ec0c84ae2c.png">
Making `supportsColumnar` as `lazy val` doesn't help as well because `supportsColumnar` is invoked twice for different objects.

Reply:
I think it is not nice that we construct some classes twice when it is not necessary. WDYT? /cc @cloud-fan @dongjoon-hyun

Reply:
Yea we shouldn't instantiate twice, but not a big problem. I'm more worried about we instantiate it for every partition.

Reply:
@MaxGekk, even if we fix this, it will still show the warning twice for schema inference and reading path at the very least. It's okay as long as we show the warning and document. Let's just go simple in this PR. This warning will be removed very soon, too.